### PR TITLE
fix(agent): apply filter_tools_regex to runtime tools

### LINKED ISSUE #4184: [Bug]: Runtime MCP tools bypass Agent.filter_tools_regex
### Is there an existing issue for the same bug?

- [x] I have searched existing issues and this is not a duplicate.

### Bug Description

`Agent.filter_tools_regex` is not applied to MCP tools materialized at conversation runtime.

Since MCP tool materialization was moved from `AgentBase._initialize()` to `LocalConversation._ensure_agent_ready()`, the current initialization flow is:

1. `AgentBase._initialize()` resolves and filters only statically configured `Agent.tools`.
2. `LocalConversation` materializes MCP tools.
3. The MCP tools are passed directly to `AgentBase.add_runtime_tools()`.
4. `add_runtime_tools()` inserts them into `_tools` without applying `filter_tools_regex`.

Although `Agent.init_state()` later calls `_initialize()` again, `_initialize()` returns immediately because `_initialized` is already `True`.

This is inconsistent with the `filter_tools_regex` field description, which states that the filter is applied after both explicitly configured tools and MCP tools are added.

This is more than a prompt-size issue: applications using the regex as a sub-agent tool boundary may expose filtered-out MCP tools to the model and allow those tools to be executed.

### Expected Behavior

`filter_tools_regex` should apply consistently to all non-default tools exposed to an Agent, including:

- Statically configured tools;
- MCP tools materialized when the conversation starts;
- MCP tools added by plugins at runtime;
- Any MCP tools dynamically advertised in the future.

An MCP tool whose name does not match the Agent's `filter_tools_regex` should:

- Not be added to `tools_map`;
- Not be sent to the LLM;
- Not be executable through that Agent.

Built-in default tools may continue to follow their existing documented behavior and remain outside this filter.

### Actual Behavior

The regex is applied only to tools resolved from `Agent.tools` during the first call to `AgentBase._initialize()`.

MCP tools are materialized afterward and passed to `add_runtime_tools()`, which adds every supplied tool directly to `_tools` without applying the regex.

As a result, MCP tools excluded by `filter_tools_regex`:

- Still appear in `agent.tools_map`;
- Are sent to the LLM;
- Can be executed by the Agent.

### Steps to Reproduce

1. Install `openhands-sdk==1.36.1`.
2. Create an Agent with `filter_tools_regex=r"^allowed$"`.
3. Initialize the Agent.
4. Add two runtime tools named `allowed` and `blocked`, reproducing the same `add_runtime_tools()` path used by `LocalConversation` for MCP tools.
5. Inspect `agent.tools_map`.
6. Observe that both `allowed` and `blocked` are present.

The production MCP path exhibiting the same behavior is:

```python
self.agent._initialize(self._state)
runtime_mcp_tools = self._runtime_mcp_tools_for_agent()
self.agent.add_runtime_tools(runtime_mcp_tools)
self.agent.init_state(...)
```

The final init_state() call does not reapply filtering because _initialize() immediately returns when _initialized is already True.

### Installation Method

Installed from PyPI using uv with `openhands-sdk==1.36.1` pinned in `pyproject.toml`.

### If you selected "Other", please specify

_No response_

### SDK Version

1.36.1

### Version Confirmation

- [x] I have confirmed this bug exists on the LATEST version of OpenHands SDK

### Python Version

3.12.3

### Model Name (if applicable)

Not model-specific. The issue occurs before the LLM request is made.

### Operating System

Linux

### Logs and Error Messages

There is no exception. The failure is observable in the resulting tool map.

With a filter that only allows `allowed`, the runtime result is:

```text
Loaded 0 tools from spec
Filtered to 0 tools after applying regex filter
['allowed', 'blocked']
```
Relevant SDK flow:

AgentBase._initialize()
    -> applies filter_tools_regex to static tools
    -> sets _initialized = True

LocalConversation._ensure_agent_ready()
    -> materializes MCP tools
    -> add_runtime_tools(runtime_mcp_tools)
    -> init_state()
       -> _initialize()
          -> returns immediately because _initialized is True

### Minimal Code Sample

  ```python
  from unittest.mock import Mock, patch

  from openhands.sdk import Agent, LLM
  from openhands.sdk.conversation.state import ConversationState
  from openhands.sdk.tool import ToolDefinition


  llm = LLM(
      model="openai/test",
      api_key="test",
  )

  agent = Agent(
      llm=llm,
      tools=[],
      include_default_tools=[],
      filter_tools_regex=r"^allowed$",
  )

  state = Mock(spec=ConversationState)

  # Avoid unrelated local vision-profile discovery during this minimal test.
  with patch(
      "openhands.sdk.agent.base.has_vision_profile_available",
      return_value=False,
  ):
      agent._initialize(state)

  allowed = Mock(spec=ToolDefinition)
  allowed.name = "allowed"

  blocked = Mock(spec=ToolDefinition)
  blocked.name = "blocked"

  # This is the same Agent API used by LocalConversation to inject
  # runtime-materialized MCP tools.
  agent.add_runtime_tools([allowed, blocked])

  print(sorted(agent.tools_map))
  assert "allowed" in agent.tools_map
  assert "blocked" not in agent.tools_map
```

Actual output:

['allowed', 'blocked']

The final assertion fails.

### Screenshots and Additional Context

_No response_

### PR DESCRIPTION
**Pull request by VascoSch92**
HUMAN:

Fixing the filter_tools_regext for runtime tools.

---

AGENT:

## Why

`Agent.filter_tools_regex` is documented as being "applied after any tools provided in `tools` and any MCP tools are added", but runtime-materialized MCP tools bypass it entirely. Tools excluded by the regex still land in `agent.tools_map`, are advertised to the LLM (`agent.py`, `tools=list(self.tools_map.values())`), and are executable — so applications using the regex as a tool boundary silently expose filtered-out MCP tools.

This is a regression: until #3964, `AgentBase._initialize()` materialized MCP tools itself and applied the regex afterwards (see `create_mcp_tools` followed by the `filter_tools_regex` block at `fe924f705^`). #3964 moved materialization into `LocalConversation._ensure_agent_ready()`, which registers the tools via `add_runtime_tools()` — a method introduced in #3826 that never applied the filter. The subsequent `init_state()` call does not re-filter because `_initialize()` returns immediately once `_initialized` is set.

## Summary

- Apply `filter_tools_regex` in `AgentBase.add_runtime_tools()`, so the single choke point covers both runtime-tool paths (conversation-start MCP tools and plugin-added tools) as well as any future dynamic-tool source.
- Built-in default tools (e.g. `InvokeSkillTool`, which the plugin path routes through the same method) are exempt, matching the existing exemption in `_initialize()`.
- Add regression tests (12) covering both paths plus edge cases — mixed batches, all-filtered batches, prefix-match anchoring, built-in subclasses, and duplicate handling around the filter; `filter_tools_regex` previously had zero test coverage anywhere in the repo, which is how this slipped through.

## Issue Number

Fixes #4184

## How to Test

End-to-end through the real production path (not just unit tests): build a `LocalConversation` whose injected `mcp_tool_provider` advertises tools named `allowed` and `blocked`, with the agent configured with `filter_tools_regex=r"^allowed$"`, then trigger `_ensure_agent_ready()` and inspect `tools_map`.

Before this fix (at `cae92b879`):

```text
filter_tools_regex = '^allowed$'
agent.tools_map    = ['allowed', 'blocked', 'inspect_image_with_vision']
AssertionError: BUG REPRODUCED: 'blocked' does not match ^allowed$ but is exposed to the agent/LLM
```

After this fix:

```text
INFO  Filtered runtime tools from 2 to 1 after applying regex filter
filter_tools_regex = '^allowed$'
agent.tools_map    = ['allowed', 'inspect_image_with_vision']
OK: filter applied to runtime MCP tools
```

(`inspect_image_with_vision` is an auto-attached built-in default tool, documented as exempt from the filter.)

That scenario is captured as `test_runtime_mcp_tools_apply_filter_tools_regex` in the new test file, so reviewers can reproduce with:

```bash
uv run pytest tests/sdk/agent/test_filter_tools_regex.py -q
```

Verification performed:

- `uv run pytest tests/sdk/agent/test_filter_tools_regex.py -q` → 12 passed. With `base.py` reverted to `origin/main`, the 6 regression tests fail and the 6 behavior-preservation tests still pass — confirming the tests catch the bug and guard existing behavior.
- `uv run pytest tests/sdk -q` (single process) → 5443 passed, 9 skipped, 13 xfailed; matches the pre-change baseline plus the 12 new tests.
- `uv run pre-commit run --files openhands-sdk/openhands/sdk/agent/base.py tests/sdk/agent/test_filter_tools_regex.py` → all hooks pass (ruff format/lint, pycodestyle, pyright, import rules, tool registration).

## Video/Screenshots

Terminal logs above (before/after output of the end-to-end reproduction and test runs).

## Type

- [x] Bug fix
- [ ] Feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Docs / chore

## Notes

- Filtering exempts built-ins via `isinstance` against `BUILT_IN_TOOL_CLASSES`, mirroring `_initialize()`'s exemption; without it, a restrictive regex would silently drop `InvokeSkillTool` injected by the plugin path and break skill invocation.
- `pattern.match` (prefix semantics) is deliberately kept consistent with `_initialize()`.
- Filtered-out MCP tools are dropped without closing their executors — identical to the pre-#3964 behavior, where `_initialize()` also discarded filtered MCP tools without cleanup. If a server's tools are all filtered its client connection lingers; flagging as a possible follow-up rather than expanding this fix.
- The plugin path has bypassed the filter since #3826 (two weeks before #3964 regressed the main `mcp_config` path); this fix covers both.

### CODE REVIEW

File: openhands-sdk/openhands/sdk/agent/base.py [hunk 1]
```diff
@@ -815,6 +820,24 @@ def add_runtime_tools(self, tools: Sequence[ToolDefinition]) -> None:
                     f"Got type: {type(tool)}"
                 )
 
+        if self.filter_tools_regex:
+            pattern = re.compile(self.filter_tools_regex)
+            builtin_classes = tuple(BUILT_IN_TOOL_CLASSES.values())
+            kept = [
```
Comment (VascoSch92) [inline openhands-sdk/openhands/sdk/agent/base.py]: Can we just not filter the tool? We can just measure the len before doing that for the logs

File: tests/sdk/agent/test_filter_tools_regex.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
+"""Tests for `Agent.filter_tools_regex`.
+
+The regex must apply both to statically configured tools resolved during
+`_initialize()` and to tools materialized at runtime (e.g. MCP tools) that are
+registered through `add_runtime_tools()`. Built-in default tools are exempt in
+both paths.
+
+Regression tests for https://github.com/OpenHands/software-agent-sdk/issues/4184
```
Comment (VascoSch92) [inline tests/sdk/agent/test_filter_tools_regex.py]: delete this. I don't want the issue

File: tests/sdk/agent/test_filter_tools_regex.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
+"""Tests for `Agent.filter_tools_regex`.
```
Comment (VascoSch92) [inline tests/sdk/agent/test_filter_tools_regex.py:1]: Is this fix well-tested? Or just happy paths?

**Review by neubig — APPROVED** [review]
Thanks!
